### ISSUE TITLE
certificates installed in /etc/ssl/certs are ignored

### ISSUE THREAD

**Issue by moscicki** [owncloud/client#1540]
Client: 1.5.2 redhat6

It looks like that sync client is not looking in all files in /etc/ssl/certs on redhat6 and hence I am getting a cert warning even though the server presents a valid cert signed by the trusted CA. Other apps are correctly finding installed certs in  /etc/ssl/certs.

Here are the cert files that client looks up at startup:

<pre>

>grep /ssl/ /tmp/trace.oc
10623 open("/etc/ssl/certs/", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 23
10623 statfs("/etc/ssl/certs/", {f_type="EXT2_SUPER_MAGIC", f_bsize=4096, f_blocks=15420929, f_bfree=3658988, f_bavail=2875628, f_files=3923968, f_ffree=3577889, f_fsid={-568994936, 1436291104}, f_namelen=255, f_frsize=4096}) = 0
10662 stat("/etc/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/lib/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/share/ssl//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/local/ssl//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/var/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/local/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10669 stat("/etc/ssl/certs//55a10908.0", 0x7f48475fba90) = -1 ENOENT (No such file or directory)
10669 stat("/usr/lib/ssl/certs//55a10908.0", 0x7f48475fba90) = -1 ENOENT (No such file or directory)
10669 stat("/usr/share/ssl//55a10908.0", 0x7f48475fba90) = -1 ENOENT (No such file or directory)
10669 stat("/usr/local/ssl//55a10908.0", 0x7f48475fba90) = -1 ENOENT (No such file or directory)
10669 stat("/var/ssl/certs//55a10908.0", 0x7f48475fba90) = -1 ENOENT (No such file or directory)
10669 stat("/usr/local/ssl/certs//55a10908.0", 0x7f48475fba90) = -1 ENOENT (No such file or directory)
10662 stat("/etc/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/lib/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/share/ssl//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/local/ssl//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/var/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/local/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/etc/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/lib/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/share/ssl//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/local/ssl//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/var/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)
10662 stat("/usr/local/ssl/certs//157753a5.0", 0x7f48552ee9c0) = -1 ENOENT (No such file or directory)

We have certificates in /etc/ssl/certs/ which are installed in various files ending in *.pem *.crt *.signing_policy ... and these bundles typically contain several certs. These files exist in addition to names which are formatted as hex-digests (e.g.  157753a5.0)

My linux support tells me that a well-behaved application should look up all files in /etc/ssl/certs and try to read the certs (or silently ignore).

I guess this is just a matter of telling Qt to do so somewhere around QSslCertificate class?

Many thanks.

**Comment by AdamWill** [comment]
@danimo "App that use OpenSSL can, by virtue of the OpenSSL API, only work with files that have a hash (created with c_rehash)."

This is not correct. OpenSSL can load either PEM-formatted bundles or directories containing individual hashed certificate symlinks. For lots and *lots* of details, see https://www.happyassassin.net/2015/01/12/a-note-about-ssltls-trusted-certificate-stores-and-platforms/ .

**Comment by danimo** [comment]
@AdamWill Yes, this was too much generalized, but not my point. OpenSSL reads from a hashed directory _or_ a single ca bundle file. I do not see where it would read independent bundles with multiple certs.

**Comment by AdamWill** [comment]
@danimo ah, I think I may have missed https://github.com/owncloud/client/issues/1540#issuecomment-49147569 - where you suggest he has some bad hash symlinks. So, Qt's SSL stuff tries to find a hashed directory and falls back on finding a bundle if it can't? Interesting...

A stock RHEL 6 should not have any hash symlinks anywhere, I don't think, so if there are some, the reporter must've caused them to exist *somehow*. But is it really finding that the symlinks exist but don't point anywhere, or is it just looking for the 'expected' hash files for the certificate being validated in common locations?

**Comment by moscicki** [comment]
@lmaestro: Yes, we still see this problem on RH6 with version 1.7.1

**Comment by moscicki** [comment]
We have fixed it in SLC6 distribution by add the cert hash file for the missing root CA file in /etc/ssl/certs

**Comment by monreal** [comment]
I am still trying to figure out how to fix this on a Fedora 21 system. I added my my cert in "BEGIN TRUSTED" format to /etc/pki/ca-trust/source/ and ran "update-ca-trust" just like /etc/pki/ca-trust/source/README says and other apps seem to pick up the cert just fine. Using strace I found that owncloud actually reads the resulting bundle file (/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem) but it does not seem to use the certificate. Any idea?
